File: Recursion/SubsetWithoutDuplicates/subset-2.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    void doRecursion(vector<vector<int>> &ans, vector<int>helper, vector<int>arr, int index){
        ans.push_back(helper);
        for(int i=index;i<arr.size();i++){
            if(i>index && arr[i]==arr[i-1]){
                continue;
            }
            helper.push_back(arr[i]);
            doRecursion(ans,helper,arr,i+1);
            helper.pop_back();
        }
    }
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int>helper;
        sort(nums.begin(),nums.end());
        doRecursion(ans,helper,nums,0);
        return ans;
    }
};

File: Recursion/SubsetWithoutDuplicates/subset-2.py
from typing import List

class Solution:
    def doRecursion(self, ans, helper,arr,index):
        ans.append(helper.copy())
        for i in range(index,len(arr)):
            if i>index and arr[i]==arr[i-1]:
                continue
            helper.append(arr[i])
            self.doRecursion(ans,helper,arr,i+1)
            helper.pop()

    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        ans = []
        nums.sort()
        self.doRecursion(ans,[],nums,0)
        return ans
        